File: cadwiki-nuget/cadwiki.AutoCAD2021.Base.Utilities/NodeGraph/NodeGraph.vb
Option Strict On
Option Infer Off
Option Explicit On
Imports Autodesk.AutoCAD.ApplicationServices
Imports Autodesk.AutoCAD.DatabaseServices
Imports Autodesk.AutoCAD.EditorInput
Imports Autodesk.AutoCAD.Geometry
Imports System.Linq

Namespace NodeGraph
    Public Class NodeGraph
        Public Nodes As New List(Of Node)
        Public Edges As New List(Of Edge)
        Public Document As Document
        Public SourceNodeId As Integer = -1
        Public DestNodeId As Integer = -2
        Public Source As Point3d = Nothing
        Public Dest As Point3d
        Public PointList As New List(Of Point3d)
        Public LayerName As String

        Public Sub New(document As Document, pointList As List(Of Point3d), destination As Point3d, source As Point3d, layerName As String)
            Me.Nodes = Nodes
            Me.Edges = Edges
            Me.Document = document
            Me.Source = source
            Me.Dest = destination
            Me.PointList = pointList
            Me.LayerName = layerName
            BuildGraph(pointList)
        End Sub

        Public Sub New(document As Document, pointList As List(Of Point3d), layerName As String)
            Me.Nodes = Nodes
            Me.Edges = Edges
            Me.Document = document
            Me.PointList = pointList
            Me.LayerName = layerName
            BuildGraph(pointList)
        End Sub

        Public Sub ModifyWithSourceAndDest(doc As Document, layerName As String, destination As Point3d, source As Point3d)
            Me.Nodes = New List(Of Node)
            Me.Edges = New List(Of Edge)

            Me.Source = source
            Me.Dest = destination
            PointList.Add(source)
            PointList.Add(destination)

            Dim filter As SelectionFilter = SelectionFilters.GetAllLineBasedEntitiesOnLayer(layerName)
            Dim graphSS As SelectionSet = SelectionSets.SelectAll(doc, filter)

            Dim closestPointOnGraphToSource As Point3d = SelectionSets.GetClosestPointOnAnyLineFromSelectionToAGivenPoint(doc, graphSS, source)
            Dim line As Line = Draw.DrawLineByPoints(doc, source, closestPointOnGraphToSource, layerName)

            Dim closestPointOnGraphToDest As Point3d = SelectionSets.GetClosestPointOnAnyLineFromSelectionToAGivenPoint(doc, graphSS, destination)
            Dim line2 As Line = Draw.DrawLineByPoints(doc, destination, closestPointOnGraphToDest, layerName)

            Dim modifiedGraph As SelectionSet = SelectionSets.SelectAll(doc, filter)
            Dim inputs As New Workflows.BreakSs.BreakSsInputs
            inputs.SelectionToBreak = modifiedGraph
            inputs.SelectionToBreakWith = modifiedGraph
            inputs.Self = True
            inputs.NewLayer = layerName
            inputs.DeleteOriginal = True
            Dim newLines As List(Of ObjectId) = Workflows.BreakSs.BreakSsWithSs(doc, inputs)

            If (Not PointList.Contains(closestPointOnGraphToSource)) Then
                PointList.Add(closestPointOnGraphToSource)
            End If

            If (Not PointList.Contains(closestPointOnGraphToDest)) Then
                PointList.Add(closestPointOnGraphToDest)
            End If


            BuildGraph(PointList)
        End Sub

        Private Sub BuildGraph(pointList As List(Of Point3d))
            Dim counter As Integer = 0
            Dim nodeId As Integer = counter
            For Each point As Point3d In pointList
                Dim parentNode As Node = Nothing
                Dim entitiesListAtNode As New List(Of Entity)
                Dim neighborList As New List(Of Node)
                Dim nodeCoordinates As Point3d = Nothing
                Dim hasNodeBeenDiscovered As Boolean = False

                nodeId = counter

                If Not Me.Source.Equals(Nothing) Then
                    Dim distanceFromSource As Double = Me.Source.DistanceTo(point)
                    If distanceFromSource = 0.0 Then
                        nodeId = SourceNodeId
                    End If
                End If

                Dim distanceFromDestination As Double = Double.MaxValue
                If Not Me.Dest.Equals(Nothing) Then
                    distanceFromDestination = Me.Dest.DistanceTo(point)
                    If distanceFromDestination = 0.0 Then
                        nodeId = DestNodeId
                    End If
                End If

                Dim node As New Node(nodeId, parentNode, entitiesListAtNode, neighborList, nodeCoordinates, hasNodeBeenDiscovered, point, distanceFromDestination)
                Nodes.Add(node)
                counter += 1
            Next
            AddNeighborsToNodes(LayerName)
        End Sub

        Public Sub AddNeighborsToNodes(layerNameToSelectFrom As String)
            Dim index As Integer = 0
            While index < Me.Nodes.Count - 1
                Dim node As Node = Me.Nodes(index)
                Dim newNode As Node = Me.AddNeighborsToNode(node, layerNameToSelectFrom)
                Me.ReplaceNode(index, newNode)
                index += 1
            End While
        End Sub

        Private Function AddNeighborsToNode(argumentNode As Node, layerNameToSelectFrom As String) As Node
            Dim entityList As List(Of Entity) = Me.GetEntitiesAtNodePoint(argumentNode, layerNameToSelectFrom)
            Dim index As Integer = 0
            While index < entityList.Count
                Dim entity As Entity = entityList(index)
                Dim line As Line = CType(entity, Line)
                Dim startPoint As Point3d = line.StartPoint
                Dim endPoint As Point3d = line.EndPoint
                Dim startNode As Node = GetNodeByPoint(startPoint)
                Dim endNode As Node = GetNodeByPoint(endPoint)
                If startNode IsNot Nothing Then
                    AddNeighborIfNotExists(argumentNode, startNode)
                End If
                If endNode IsNot Nothing Then
                    AddNeighborIfNotExists(argumentNode, endNode)
                End If
                index += 1
            End While
            Return argumentNode
        End Function

        Private Sub AddNeighborIfNotExists(argumentNode As Node, neighbor As Node)
            Dim areNodesTheSame As Integer = argumentNode.NodeId.CompareTo(neighbor.NodeId)
            'update argumentNode with neighbor
            If Not argumentNode.NeighborList.Contains(neighbor) And Not argumentNode.NodeId.Equals(neighbor.NodeId) Then
                argumentNode.NeighborList.Add(neighbor)
            End If
            'update neighbor with argumentNode
            If Not neighbor.NeighborList.Contains(argumentNode) And Not argumentNode.NodeId.Equals(neighbor.NodeId) Then
                neighbor.NeighborList.Add(argumentNode)
            End If
        End Sub

        Private Function GetNodeByPoint(point As Point3d) As Node
            For Each node As Node In Me.Nodes
                If node.AutoCADPoint.Equals(point) Then
                    Return node
                End If
            Next
            Return Nothing
        End Function
        Private Function GetEntitiesAtNodePoint(node As Node, layerNameToSelectFrom As String) As List(Of Entity)
            Dim fuzz As Double = 0.001
            Dim point As Point3d = node.AutoCADPoint
            Dim pt1 As Point3d = New Point3d(point.X + fuzz, point.Y + fuzz, point.Z)
            Dim pt2 As Point3d = New Point3d(point.X - fuzz, point.Y - fuzz, point.Z)
            Dim filter As SelectionFilter = SelectionFilters.GetAllLineEntitiesOnLayer(layerNameToSelectFrom)
            Dim ss As SelectionSet = SelectionSets.CrossingWindow(Document, pt2, pt2, filter)
            Dim entityListAtNode As List(Of Entity) = SelectionSets.GetEntityList(Document, ss)
            Return entityListAtNode
        End Function

        Private Sub ReplaceNode(index As Integer, newNode As Node)
            Me.Nodes(index) = newNode
        End Sub

        Public Sub LabelNodes()
            Dim newLine As String = vbCrLf
            For Each node As Node In Nodes
                Dim neighborIds As List(Of String) = node.GetNeighborIdsToStringList()
                Dim formattedNeighborString As String = String.Join(", ", neighborIds)

                Dim label As String = String.Format("{0}ID:{1}{2}Nbors:{3}", newLine, node.NodeId, newLine, formattedNeighborString)

                Dim settings As Mtexts.Settings = New Mtexts.Settings()
                settings.Height = 2.5
                settings.Content = label
                settings.Location = node.AutoCADPoint
                settings.Location = New Point3d(settings.Location.X + 0.1, settings.Location.Y, settings.Location.Z)

                Dim mtext As MText = Mtexts.Add(settings)

                Draw.DrawCircleAtLocation(node.AutoCADPoint, 0.1)
            Next
        End Sub


        Public Function FindNodeById(nodeId As Integer) As Node
            Return Nodes.Find(Function(n) n.NodeId = nodeId)
        End Function

        Public Function BFS(startNodeId As Integer, destinationNodeId As Integer) As List(Of Node)
            Dim visitedNodes As New HashSet(Of Node)()
            Dim queue As New Queue(Of Node)()

            Dim startNode As Node = FindNodeById(startNodeId)
            Dim destinationNode As Node = FindNodeById(destinationNodeId)

            queue.Enqueue(startNode)
            visitedNodes.Add(startNode)

            While queue.Count > 0
                Dim currentNode As Node = queue.Dequeue()

                ' Check if the destination node has been reached
                If currentNode.NodeId = destinationNodeId Then
                    ' Build and return the path from start to destination
                    Return BuildPath(currentNode)
                End If

                ' Explore neighbors of the current node
                For Each neighborNode As Node In currentNode.NeighborList
                    If Not visitedNodes.Contains(neighborNode) Then
                        queue.Enqueue(neighborNode)
                        visitedNodes.Add(neighborNode)
                        neighborNode.ParentNode = currentNode ' Set the parent node for backtracking the path
                    End If
                Next
            End While

            ' No path found
            Return Nothing
        End Function

        Private Function BuildPath(destinationNode As Node) As List(Of Node)
            Dim path As New List(Of Node)()
            Dim currentNode As Node = destinationNode

            ' Backtrack from destination to start using parent pointers
            While currentNode IsNot Nothing
                path.Insert(0, currentNode) ' Insert at the beginning to maintain the correct order
                currentNode = currentNode.ParentNode
            End While

            Return path
        End Function


        Public Sub DrawLinesAlongPath(doc As Document, path As List(Of Node), layerName As String)
            If path IsNot Nothing Then
                Try
                    Dim startPoint As Point3d = path(0).AutoCADPoint
                    For i As Integer = 1 To path.Count - 1
                        Dim endPoint As Point3d = path(i).AutoCADPoint
                        Draw.DrawLineByPoints(doc, startPoint, endPoint, layerName)
                        startPoint = endPoint
                    Next
                Catch ex As Exception
                End Try
            Else

            End If
        End Sub

    End Class
End Namespace

File: cadwiki-nuget/cadwiki.AC/NodeGraph/NodeGraph.cs
using System;
using System.Collections.Generic;
using Autodesk.AutoCAD.ApplicationServices;
using Autodesk.AutoCAD.DatabaseServices;
using Autodesk.AutoCAD.Geometry;
using Microsoft.VisualBasic;
using cadwiki.AC.Utilities;

namespace cadwiki.AC.NodeGraph
{
    public class NodeGraph
    {
        public List<Node> Nodes = new List<Node>();
        public List<Edge> Edges = new List<Edge>();
        public Document Document;
        public int SourceNodeId = -1;
        public int DestNodeId = -2;
        public Point3d Source = default;
        public Point3d Dest;
        public List<Point3d> PointList = new List<Point3d>();
        public string LayerName;

        public NodeGraph(Document document, List<Point3d> pointList, Point3d destination, Point3d source, string layerName)
        {
            Nodes = Nodes;
            Edges = Edges;
            Document = document;
            Source = source;
            Dest = destination;
            PointList = pointList;
            LayerName = layerName;
            BuildGraph(pointList);
        }

        public NodeGraph(Document document, List<Point3d> pointList, string layerName)
        {
            Nodes = Nodes;
            Edges = Edges;
            Document = document;
            PointList = pointList;
            LayerName = layerName;
            BuildGraph(pointList);
        }

        public void ModifyWithSourceAndDest(Document doc, string layerName, Point3d destination, Point3d source)
        {
            Nodes = new List<Node>();
            Edges = new List<Edge>();

            Source = source;
            Dest = destination;
            PointList.Add(source);
            PointList.Add(destination);

            var filter = SelectionFilters.GetAllLineBasedEntitiesOnLayer(layerName);
            var graphSS = SelectionSets.SelectAll(doc, filter);

            var closestPointOnGraphToSource = SelectionSets.GetClosestPointOnAnyLineFromSelectionToAGivenPoint(doc, graphSS, source);
            var line = Draw.DrawLineByPoints(doc, source, closestPointOnGraphToSource, layerName);

            var closestPointOnGraphToDest = SelectionSets.GetClosestPointOnAnyLineFromSelectionToAGivenPoint(doc, graphSS, destination);
            var line2 = Draw.DrawLineByPoints(doc, destination, closestPointOnGraphToDest, layerName);

            var modifiedGraph = SelectionSets.SelectAll(doc, filter);
            var inputs = new Workflows.BreakSs.BreakSsInputs();
            inputs.SelectionToBreak = modifiedGraph;
            inputs.SelectionToBreakWith = modifiedGraph;
            inputs.Self = true;
            inputs.NewLayer = layerName;
            inputs.DeleteOriginal = true;
            var newLines = Workflows.BreakSs.BreakSsWithSs(doc, inputs);

            if (!PointList.Contains(closestPointOnGraphToSource))
            {
                PointList.Add(closestPointOnGraphToSource);
            }

            if (!PointList.Contains(closestPointOnGraphToDest))
            {
                PointList.Add(closestPointOnGraphToDest);
            }


            BuildGraph(PointList);
        }

        private void BuildGraph(List<Point3d> pointList)
        {
            int counter = 0;
            int nodeId = counter;
            foreach (Point3d point in pointList)
            {
                Node parentNode = null;
                var entitiesListAtNode = new List<Entity>();
                var neighborList = new List<Node>();
                Point3d nodeCoordinates = default;
                bool hasNodeBeenDiscovered = false;

                nodeId = counter;

                if (!Source.Equals(null))
                {
                    double distanceFromSource = Source.DistanceTo(point);
                    if (distanceFromSource == 0.0d)
                    {
                        nodeId = SourceNodeId;
                    }
                }

                double distanceFromDestination = double.MaxValue;
                if (!Dest.Equals(null))
                {
                    distanceFromDestination = Dest.DistanceTo(point);
                    if (distanceFromDestination == 0.0d)
                    {
                        nodeId = DestNodeId;
                    }
                }

                var node = new Node(nodeId, parentNode, entitiesListAtNode, neighborList, nodeCoordinates, hasNodeBeenDiscovered, point, distanceFromDestination);
                Nodes.Add(node);
                counter += 1;
            }
            AddNeighborsToNodes(LayerName);
        }

        public void AddNeighborsToNodes(string layerNameToSelectFrom)
        {
            int index = 0;
            while (index < Nodes.Count - 1)
            {
                var node = Nodes[index];
                var newNode = AddNeighborsToNode(node, layerNameToSelectFrom);
                ReplaceNode(index, newNode);
                index += 1;
            }
        }

        private Node AddNeighborsToNode(Node argumentNode, string layerNameToSelectFrom)
        {
            var entityList = GetEntitiesAtNodePoint(argumentNode, layerNameToSelectFrom);
            int index = 0;
            while (index < entityList.Count)
            {
                var entity = entityList[index];
                Line line = (Line)entity;
                var startPoint = line.StartPoint;
                var endPoint = line.EndPoint;
                var startNode = GetNodeByPoint(startPoint);
                var endNode = GetNodeByPoint(endPoint);
                if (startNode is not null)
                {
                    AddNeighborIfNotExists(argumentNode, startNode);
                }
                if (endNode is not null)
                {
                    AddNeighborIfNotExists(argumentNode, endNode);
                }
                index += 1;
            }
            return argumentNode;
        }

        private void AddNeighborIfNotExists(Node argumentNode, Node neighbor)
        {
            int areNodesTheSame = argumentNode.NodeId.CompareTo(neighbor.NodeId);
            // update argumentNode with neighbor
            if (!argumentNode.NeighborList.Contains(neighbor) & !argumentNode.NodeId.Equals(neighbor.NodeId))
            {
                argumentNode.NeighborList.Add(neighbor);
            }
            // update neighbor with argumentNode
            if (!neighbor.NeighborList.Contains(argumentNode) & !argumentNode.NodeId.Equals(neighbor.NodeId))
            {
                neighbor.NeighborList.Add(argumentNode);
            }
        }

        private Node GetNodeByPoint(Point3d point)
        {
            foreach (Node node in Nodes)
            {
                if (node.AutoCADPoint.Equals(point))
                {
                    return node;
                }
            }
            return null;
        }
        private List<Entity> GetEntitiesAtNodePoint(Node node, string layerNameToSelectFrom)
        {
            double fuzz = 0.001d;
            var point = node.AutoCADPoint;
            var pt1 = new Point3d(point.X + fuzz, point.Y + fuzz, point.Z);
            var pt2 = new Point3d(point.X - fuzz, point.Y - fuzz, point.Z);
            var filter = SelectionFilters.GetAllLineEntitiesOnLayer(layerNameToSelectFrom);
            var ss = SelectionSets.CrossingWindow(Document, pt2, pt2, filter);
            var entityListAtNode = SelectionSets.GetEntityList(Document, ss);
            return entityListAtNode;
        }

        private void ReplaceNode(int index, Node newNode)
        {
            Nodes[index] = newNode;
        }

        public void LabelNodes()
        {
            string newLine = Constants.vbCrLf;
            foreach (Node node in Nodes)
            {
                var neighborIds = node.GetNeighborIdsToStringList();
                string formattedNeighborString = string.Join(", ", neighborIds);

                string label = string.Format("{0}ID:{1}{2}Nbors:{3}", newLine, node.NodeId, newLine, formattedNeighborString);

                var settings = new Mtexts.Settings();
                settings.Height = 2.5d;
                settings.Content = label;
                settings.Location = node.AutoCADPoint;
                settings.Location = new Point3d(settings.Location.X + 0.1d, settings.Location.Y, settings.Location.Z);

                var mtext = Mtexts.Add(settings);

                Draw.DrawCircleAtLocation(node.AutoCADPoint, 0.1d);
            }
        }


        public Node FindNodeById(int nodeId)
        {
            return Nodes.Find(n => n.NodeId == nodeId);
        }

        public List<Node> BFS(int startNodeId, int destinationNodeId)
        {
            var visitedNodes = new HashSet<Node>();
            var queue = new Queue<Node>();

            var startNode = FindNodeById(startNodeId);
            var destinationNode = FindNodeById(destinationNodeId);

            queue.Enqueue(startNode);
            visitedNodes.Add(startNode);

            while (queue.Count > 0)
            {
                var currentNode = queue.Dequeue();

                // Check if the destination node has been reached
                if (currentNode.NodeId == destinationNodeId)
                {
                    // Build and return the path from start to destination
                    return BuildPath(currentNode);
                }

                // Explore neighbors of the current node
                foreach (Node neighborNode in currentNode.NeighborList)
                {
                    if (!visitedNodes.Contains(neighborNode))
                    {
                        queue.Enqueue(neighborNode);
                        visitedNodes.Add(neighborNode);
                        neighborNode.ParentNode = currentNode; // Set the parent node for backtracking the path
                    }
                }
            }

            // No path found
            return null;
        }

        private List<Node> BuildPath(Node destinationNode)
        {
            var path = new List<Node>();
            var currentNode = destinationNode;

            // Backtrack from destination to start using parent pointers
            while (currentNode is not null)
            {
                path.Insert(0, currentNode); // Insert at the beginning to maintain the correct order
                currentNode = currentNode.ParentNode;
            }

            return path;
        }


        public void DrawLinesAlongPath(Document doc, List<Node> path, string layerName)
        {
            if (path is not null)
            {
                try
                {
                    var startPoint = path[0].AutoCADPoint;
                    for (int i = 1, loopTo = path.Count - 1; i <= loopTo; i++)
                    {
                        var endPoint = path[i].AutoCADPoint;
                        Draw.DrawLineByPoints(doc, startPoint, endPoint, layerName);
                        startPoint = endPoint;
                    }
                }
                catch (Exception ex)
                {
                }
            }
            else
            {

            }
        }

    }
}